pkg/auth/dependency/webapp/authenticate_provider_impl.go

```diff
@@ -410,6 +382,11 @@ func (p *AuthenticateProviderImpl) LinkIdentityProvider(w http.ResponseWriter, r
 
 	userID := auth.GetSession(r.Context()).AuthnAttrs().UserID
 
+	_, err = p.restoreState(r)
```

Comment:
I suppose this means an expired/invalid SID would fail the POST request and generate a new SID for the next request?

I think we need to reconsider whether to automatically re-generate the state in `persistState`, since now it has many important data that cannot be recovered if lost (e.g. interaction token, redirect URI).

Reply:
We have a state middleware to detect expired/invalid SID and redirect the user to root if it is the case. Of course if we read the state twice we could have data race, e.g. the second read is considered expired.

Reply:
I suppose we can just read the state from store and inject it into context in middleware, so that `restoreState` would just read from the context and `persistState` would update state in context and write to store. Therefore, we would not re-generate expired/invalid SID anymore.

Reply:
Let me open [a new issue](https://github.com/SkygearIO/skygear-server/issues/1409) to address this to avoid dragging this PR.